[SPARK-15704] [SQL] add a test case in DatasetAggregatorSuite for regression testing

inouehrs opened the pull request:
## What changes were proposed in this pull request?

This change fixes a crash in TungstenAggregate while executing "Dataset complex Aggregator" test case due to IndexOutOfBoundsException.

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-15704
## How was this patch tested?

Using existing unit tests (including DatasetBenchmark)

inouehrs committed:
fix a crash in TangstenAggregate while executing DatasetBenchmark

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
cc @cloud-fan / @sameeragarwal

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #59853 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59853/consoleFull)** for PR 13446 at commit [`12deb23`](https://github.com/apache/spark/commit/12deb235ba5e3a33b3e7a0df6899ffbfa57d14fa).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala:
```diff
@@ -34,6 +34,11 @@ object TypedAggregateExpression {
     val bufferEncoder = encoderFor[BUF]
     val bufferSerializer = bufferEncoder.namedExpressions
     val bufferDeserializer = bufferEncoder.deserializer.transform {
+      case UnresolvedAttribute(nameParts) =>
```
good catch! can you try if we can fix it in this way? https://github.com/apache/spark/pull/13269/files#diff-585631b931d2b881e65c585dec11cadbR36

I think it's less hacky

inouehrs replied:
It seems that the new cleaner implementation does not need my change; it handles unresolved attributes correctly. Thank you!

cloud-fan replied:
we can apply that part of code in this PR and fix this bug first

SparkQA commented:
**[Test build #59853 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59853/consoleFull)** for PR 13446 at commit [`12deb23`](https://github.com/apache/spark/commit/12deb235ba5e3a33b3e7a0df6899ffbfa57d14fa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59853/
Test PASSed.

rxin commented:
Thanks - can you also update the title? TangstenAggregate -> TungstenAggregate

inouehrs commented:
I fixed my typos.

inouehrs committed:
rewrite based on discussion

https://github.com/apache/spark/pull/13446#discussion_r65575752

SparkQA commented:
**[Test build #59919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59919/consoleFull)** for PR 13446 at commit [`65a1534`](https://github.com/apache/spark/commit/65a1534c101f1bc0dd711af88d08dedc42df6c08).

inouehrs committed:
add a test case based on DatasetBenchmark

This test case causes IndexOutOfBoundsException without fix in
TypedAggregateExpression

inouehrs commented:
I rewrote the patch as suggested by @cloud-fan (based on https://github.com/apache/spark/pull/13269/files#diff-585631b931d2b881e65c585dec11cadbR36).
Also, I added a test case to reproduce the error in DatasetAggregatorSuite.

SparkQA commented:
**[Test build #59924 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59924/consoleFull)** for PR 13446 at commit [`3fb1dd3`](https://github.com/apache/spark/commit/3fb1dd3bdaeebcea7f832aaa1fdbdfcac53278a5).

SparkQA commented:
**[Test build #59919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59919/consoleFull)** for PR 13446 at commit [`65a1534`](https://github.com/apache/spark/commit/65a1534c101f1bc0dd711af88d08dedc42df6c08).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59919/
Test PASSed.

SparkQA commented:
**[Test build #59924 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59924/consoleFull)** for PR 13446 at commit [`3fb1dd3`](https://github.com/apache/spark/commit/3fb1dd3bdaeebcea7f832aaa1fdbdfcac53278a5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59924/
Test PASSed.

inouehrs committed:
Merge branch 'apache/master' into fix_aggregate

SparkQA commented:
**[Test build #59936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59936/consoleFull)** for PR 13446 at commit [`3837825`](https://github.com/apache/spark/commit/3837825e317ba459fc1824dd0e7ae1c63a25a07b).

inouehrs commented:
Since the new encoder framework ([SPARK-15494]) has been merged, my bug fix part is no longer needed.
I appreciate if the new test case part can be merged. Thanks!

cloud-fan commented:
can you revert the file mode change? the test LGTM

inouehrs committed:
revert unintended file mode change

inouehrs commented:
I reverted the file mode change. Thank you!

SparkQA commented:
**[Test build #59939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59939/consoleFull)** for PR 13446 at commit [`e892324`](https://github.com/apache/spark/commit/e8923249c651624526136891a75eae725f316d60).

SparkQA commented:
**[Test build #59936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59936/consoleFull)** for PR 13446 at commit [`3837825`](https://github.com/apache/spark/commit/3837825e317ba459fc1824dd0e7ae1c63a25a07b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59936/
Test PASSed.

SparkQA commented:
**[Test build #59939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59939/consoleFull)** for PR 13446 at commit [`e892324`](https://github.com/apache/spark/commit/e8923249c651624526136891a75eae725f316d60).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59939/
Test PASSed.

cloud-fan commented:
can you also update your title to say that this PR only adds regression test? thanks!

inouehrs commented:
TItle changed. Thanks

rxin commented:
@cloud-fan next time please leave a message on the pr saying it was merged and the branches it was merged in.

cloud-fan commented:
Sorry I was interrupted by something and forgot about it...
thanks, merging to master and 2.0!
